golem_core/storage/gftp.py

```diff
@@ -0,0 +1,456 @@
+
+
+class GftpSource(Source):
+    def __init__(self, length: int, link: PubLink):
+        self._len = length
+        self._link = link
+
+    @property
+    def download_url(self) -> str:
+        return self._link["url"]
+
+    @property
+    def path(self) -> Path:
+        return Path(self._link["file"])
+
+    async def content_length(self) -> int:
+        return self._len
+
+
+class GftpDestination(Destination):
+    def __init__(self, _proc: GftpDriver, _link: PubLink) -> None:
+        self._proc = _proc
+        self._link = _link
+
+    @property
+    def upload_url(self) -> str:
+        return self._link["url"]
+
+    async def download_stream(self) -> Content:
+        file_path = Path(self._link["file"])
+        length = file_path.stat().st_size
+
+        async def chunks() -> AsyncIterator[bytes]:
+            with open(file_path, "rb") as f:
+                chunk = f.read(30_000)
+                while chunk:
+                    yield chunk
+                    chunk = f.read(30_000)
+
+        return Content(length=length, stream=chunks())
+
+    async def download_file(self, destination_file: PathLike) -> None:
+        if str(destination_file) == self._link["file"]:
+            return
+        return await super().download_file(destination_file)
+
+
+@contextlib.contextmanager
+def _temp_file(temp_dir: Path) -> Iterator[Path]:
+    """Create a new temporary file in `temp_dir`.
+
+    Implements the ContextManager interface. Deletes the file on ContextManager's exit.
+    """
+    fd, name = tempfile.mkstemp(prefix="yapapi_", dir=temp_dir)
+    os.close(fd)
+    path = Path(name)
+    try:
+        yield path
+    finally:
+        _delete_if_exists(path)
+
+
+def _delete_if_exists(path: Path) -> None:
+    if path.exists():
+        try:
+            path.unlink()
+            _logger.debug("Deleted temporary file %s", path)
+        except PermissionError as err:
+            # We're on Windows and using `gftp` < 0.7.3, so the file is kept open
+            # by `gftp` and cannot be deleted.
+            _logger.debug("Cannot delete file: %s", err)
+
+
+USE_GFTP_CLOSE_ENV_VAR = "YAPAPI_USE_GFTP_CLOSE"
+"""The environment variable used by GftpProvider to control whether `gftp close` should be used."""
+
+
+def read_use_gftp_close_env_var() -> Optional[bool]:
+    """Determine from the environment whether `GftpProvider` should use the `gftp close` command.
+
+    Reads the value of the environment variable with the name stored in `USE_GFTP_CLOSE_ENV_VAR`.
+    If the environment variable is set and its value can be interpreted as boolean value by
+    `distutils.util.strtobool()` then the corresponding boolean value is returned. Otherwise,
+    returns `None`.
+    """
+    try:
+        env_value = os.environ.get(USE_GFTP_CLOSE_ENV_VAR, "")
+        return bool(strtobool(env_value))
+    except ValueError:
+        return None
+
+
+MIN_GFTP_VERSION_THAT_CAN_GFTP_CLOSE = semantic_version.Version("0.7.3")
```

Comment:
I think this will be 2 year ago soon? Maybe this is a good moment for some cleanup here? Not insisting though

Reply:
I will add a separated task for it. I don't want to overcharge this class before we integrate yapapi with golem-core.